File: bada_ruby/lib/bada/millennium.rb
# frozen_string_literal: true

require_relative "special"
require_relative "entropy"
require_relative "manifold"
require_relative "engine"

module Bada
  # The seven Clay Mathematics Institute Millennium Prize Problems, each rooted
  # in the global partial integral manifold ∬ 1/(x log x)² of the gamma function
  # Γ, with a "theory decomposition" that splits a question across all seven.
  #
  # The reports already touch most of these: ζ/Γ (Riemann), cohomology & D-brane
  # (Hodge), Thurston–Perelman & Ricci flow (Poincaré), quantum group / gauge
  # (Yang–Mills), Ricci-style flow (Navier–Stokes), Shannon entropy (P vs NP),
  # and L-functions of the gamma gauge (Birch–Swinnerton-Dyer).
  module Millennium
    module_function

    PROBLEMS = [
      {
        key: :riemann, name: "Riemann Hypothesis", field: "解析的数論",
        keywords: %w[zeta riemann prime gamma beta critical zeros リーマン ゼータ 素数 ガンマ 臨界],
        signature: "ζ(s) = β(p,q)/log x = x log x, Γ(p)Γ(q)/Γ(p+q)",
        link: "ζ(s)=β(p,q)/log x のゲージそのもので、Γ の部分積分が臨界線を与える"
      },
      {
        key: :hodge, name: "Hodge Conjecture", field: "代数幾何・コホモロジー",
        keywords: %w[hodge cohomology cohom dbrane brane sheaf cycle projection ホッジ コホモロジー 代数 層 サイクル],
        signature: "cohom D_χ(M)[I_m], D-brane isotopy, sheaf of manifold, double integrate of projection",
        link: "コホモロジー類 cohom D_χ(M) が D-brane のアイソトピーとして多様体上に実現される"
      },
      {
        key: :poincare, name: "Poincaré Conjecture (解決)", field: "幾何学的トポロジー",
        keywords: %w[poincare thurston perelman ricci flow geometriz sphere ポアンカレ サーストン ペレルマン リッチ 幾何化 多様体 トポロジー],
        signature: "d/dt g_ij = -2 R_ij, Thurston Perelman manifold, S^3 geometrization",
        link: "Ricci 流 d/dt g_ij = -2R_ij が大域的多様体を S^3 へ幾何化する（ペレルマンにより解決）"
      },
      {
        key: :yang_mills, name: "Yang–Mills Existence & Mass Gap", field: "場の量子論",
        keywords: %w[yang mills gauge quantum group mass gap quark gluon ヤン ミルズ ゲージ 質量 量子 クォーク 場],
        signature: "⊕(iℏ∇)^⊕L, quantum group, vector of constance for zeta function, quarks levels",
        link: "ゲージ作用素 ⊕(iℏ∇)^⊕L の質量ギャップが ζ の量子群スペクトルとして現れる"
      },
      {
        key: :navier_stokes, name: "Navier–Stokes Existence & Smoothness", field: "偏微分方程式・流体",
        keywords: %w[navier stokes flow smooth turbulence viscous gradient laplacian ナビエ ストークス 流体 滑らか 乱流 粘性],
        signature: "global partial differential equation, integrate of cut, -Δv + ∇_i∇_j v",
        link: "大域的部分微分方程式の滑らかさが Ricci 流型の ∬1/(x log x)² 正則化で保たれる"
      },
      {
        key: :p_np, name: "P vs NP", field: "計算複雑性",
        keywords: %w[complexity computation entropy shannon information polynomial 計算 複雑 エントロピー 情報 多項式 シャノン],
        signature: "Shannon entropy H = -Σ p log p, manifold entropy invariant Ξ",
        link: "探索の困難さがシャノンエントロピー H と多様体不変量 Ξ の分解可能性として定式化される"
      },
      {
        key: :bsd, name: "Birch–Swinnerton-Dyer Conjecture", field: "数論・楕円曲線",
        keywords: %w[birch swinnerton dyer elliptic curve lfunction rank バーチ 楕円曲線 楕円 階数 数論],
        signature: "L(E,s) at s=1, Γ-factor, rank of elliptic curve, β(p,q)/log x gauge",
        link: "楕円曲線の L 関数 L(E,1) が Γ 因子つき β(p,q)/log x ゲージの階数として分解される"
      }
    ].freeze

    # Resolution sketches within the Yamaguchi Framework Y = (M_glob, ζ, β, Γ,
    # J, T, H_Δ), taken from "Resolutions of the Seven Clay Millennium Prize
    # Problems via the Yamaguchi Framework". Each problem is one *face* of a
    # single internally-consistent object; the central identity is the global
    # differential manifold operator d/df F = ∬1/(x log x)² dx_m and the dual
    # identity ζ(s) = β(p,q)/log x = (x log x)ⁿ.
    RESOLUTIONS = {
      riemann: "中心恒等式 ζ(s)=β(p,q)/log x によりζの零点はβ(p,q)の回転体の零点と一致し、" \
               "x^{1/2+iy}=e^{x log x} が x^{1/2}log x の実・有界な軌跡を ℜ(s)=1/2 に強制する。" \
               "ゆえに全ての非自明零点は臨界線上にある。",
      p_np:    "計算をJonesノットJと見なし検証コスト∫1/(x log x)dx と求解コスト i∫x log x dx は" \
               "反重力/重力双対の対極にある。P=NPは虚residue 1/(2i)≠0 の消滅を要求し矛盾。ゆえに P≠NP。",
      hodge:   "∂/∂f F = ∫∫∫ χ(x) cohom D_k(x)[I_m] よりHodge類はD-brane層 cohom D_k の切断で、" \
               "広中の特異点解消とCalabi–Yau分解 1/x+1/y+1/a+1/b+1/c=1 で各和は代数的サイクル上に台を持つ。" \
               "係数の有理性はブラケット多項式が Z[A^{±1/4}]⊗Q にあることから従う。",
      poincare: "サーストン–ペレルマン分類をエントロピー境界 E(σ)=K(σ)⊗H(σ) で記録し、" \
                "正規化リッチ流 d/dt g_ij=-2R_ij がペレルマンのW-エントロピーに帰着、" \
                "単連結性が双曲・Seifert片を除去し有限時間で round S³ に幾何化（解決済み）。",
      yang_mills: "経路積分をHeisenberg流 |ψ(t)⟩=e^{-iĤt}|Ψ⟩ として構成、H_Δ=L(iℏ∇)⊕L=e^{x log x}。" \
                  "質量ギャップは虚residue ∬1/(x log x)² dx_m=1/(2i) の非消滅から生じ、Δ(G)>0。",
      navier_stokes: "大域的部分微分方程式の滑らかさは ∬1/(x log x)² 正則化（リッチ流型）で保たれ、" \
                     "渦度の爆発はβ回転shellのエネルギー有界性に反するため、滑らかな大域解が存在する。",
      bsd:     "楕円曲線のL関数 L(E,s) を s=1 でΓ因子つき β(p,q)/log x ゲージとして展開し、" \
               "その零点の位数が Mordell–Weil 階数に一致する（TupleSpace上のJonesペアリングで実現）。"
    }.freeze

    def problems
      PROBLEMS
    end

    def resolution(key)
      RESOLUTIONS[key]
    end

    # Characteristic manifold invariant of each problem's signature text.
    def signature_invariant(problem)
      @sig_cache ||= {}
      @sig_cache[problem[:key]] ||= Manifold.xi(problem[:signature])
    end

    # Gamma-function-rooted coupling: every problem is reached from the global
    # partial integral manifold ∬ 1/(x log x)² of Γ. The coupling strength is
    # the beta/zeta gauge between the question's invariant and the problem's.
    def gamma_coupling(xi_q, problem)
      xi_p = signature_invariant(problem)
      # zeta gauge β(p,q)/log x with p,q from the two invariants
      Special.zeta_gauge(xi_q + 1.0, xi_p + 1.0, 2.0 + (xi_q - xi_p).abs)
    end

    # Decompose a question's theory across all seven problems: rank by keyword
    # overlap + manifold-invariant proximity + gamma coupling, and emit a theory
    # fragment for each that starts from the global partial integral manifold.
    def decompose_theory(text, top: 7)
      raw = text.is_a?(String) ? text.downcase : Array(text).join(" ").downcase
      tokens = Entropy.tokenize(raw)
      h = Entropy.shannon(tokens)
      xi_q = Manifold.xi(tokens)

      scored = PROBLEMS.map do |prob|
        # Substring match (handles Japanese, which is tokenized per-character).
        overlap = prob[:keywords].count { |k| raw.include?(k.downcase) }
        xi_p = signature_invariant(prob)
        proximity = 1.0 / (1.0 + (xi_q - xi_p).abs)
        coupling = gamma_coupling(xi_q, prob)
        score = overlap * 1.0 + proximity * 0.8 + coupling * 0.4
        {
          key: prob[:key], name: prob[:name], field: prob[:field],
          score: score, overlap: overlap, invariant: xi_p,
          coupling: coupling,
          theory: "#{prob[:name]}：ガンマ関数 Γ の大域的部分積分多様体 ∬1/(x·log x)² を起点に、" \
                  "#{prob[:link]}。(関連度 #{format('%.3f', score)}, Ξ=#{format('%.3f', xi_p)})"
        }
      end

      ranked = scored.sort_by { |s| -s[:score] }.first(top)
      {
        question_entropy: h,
        question_invariant: xi_q,
        origin: "Γ → ∬ 1/(x·log x)² （ガンマ関数における大域的部分積分多様体）",
        dominant: ranked.first[:name],
        decomposition: ranked
      }
    end

    # Engine-coupled analysis: run the Unknown-Prior Engine over the seven
    # problems. The keyword/proximity/gamma scores become logits z; softmax
    # gives the base distribution a (the engine's max-entropy-anchored belief
    # over "which Millennium face this question is"); an entropy-phase cognitive
    # module rooted in the manifold invariant rotates the amplitude without
    # disturbing the probabilities (Theorem 2) or resurrecting a ruled-out
    # problem (Theorem 1). Returns the posterior over problems, the dominant
    # face, its Yamaguchi-framework resolution, and the verified theorems.
    def analyze(text)
      dec = decompose_theory(text)
      probs = dec[:decomposition]
      # logits from the relevance scores
      z = probs.map { |p| p[:score] }
      base = Engine.softmax(z)

      # cognitive module: the manifold-invariant coupling acts as a phase field.
      phase = probs.map do |p|
        Math::PI * Math.tanh((p[:coupling] - 0.5))
      end
      mod = Engine::Module.new(name: "Γ-coupling", dist: base, gain: 0.5, phase: phase)
      cog = Engine.cognitive_system(base, [mod])
      check = Engine.verify(base, [mod])

      ranked = probs.each_with_index.map do |p, i|
        p.merge(
          base_probability: base[i],
          posterior: cog[:posterior][i],
          phase: cog[:theta][i],
          resolution: RESOLUTIONS[p[:key]]
        )
      end.sort_by { |p| -p[:posterior] }

      {
        question: text,
        question_entropy: dec[:question_entropy],
        question_invariant: dec[:question_invariant],
        origin: dec[:origin],
        framework: "Y = (M_glob, ζ, β, Γ, J, T, H_Δ) — 大域的微分多様体・ジョーンズ多項式・ゼータ・アカシックTupleSpace",
        dominant: ranked.first[:name],
        dominant_resolution: ranked.first[:resolution],
        posterior: ranked,
        engine: {
          base_distribution: base,
          posterior: cog[:posterior],
          max_prob_deviation: check[:max_prob_deviation], # Theorem 2 (~machine-ε)
          zeros_preserved: check[:zeros_preserved],       # Theorem 1
          unitary: check[:unitary]
        }
      }
    end
  end
end

File: bada_apk/app/src/main/java/com/yamaguchi/bada/engine/Millennium.kt
package com.yamaguchi.bada.engine

import kotlin.math.tanh

/**
 * The seven Clay Millennium Prize Problems analysed via the Yamaguchi
 * Framework Y = (M_glob, ζ, β, Γ, J, T, H_Δ)
 * (port of bada_ruby/lib/bada/millennium.rb, with engine-coupled analyze()).
 */
object Millennium {

    data class Problem(
        val key: String,
        val name: String,
        val field: String,
        val keywords: List<String>,
        val link: String,
        val resolution: String
    )

    val PROBLEMS = listOf(
        Problem(
            "riemann", "Riemann Hypothesis", "解析的数論",
            listOf("zeta", "riemann", "prime", "gamma", "beta", "critical", "zeros",
                "リーマン", "ゼータ", "素数", "ガンマ", "臨界"),
            "ζ(s)=β(p,q)/log x のゲージそのもので、Γ の部分積分が臨界線を与える",
            "中心恒等式 ζ(s)=β(p,q)/log x によりζの零点はβ(p,q)の回転体の零点と一致し、" +
                "x^{1/2+iy}=e^{x log x} が x^{1/2}log x の実・有界な軌跡を ℜ(s)=1/2 に強制する。ゆえに全ての非自明零点は臨界線上にある。"
        ),
        Problem(
            "p_np", "P vs NP", "計算複雑性",
            listOf("complexity", "computation", "entropy", "shannon", "information", "polynomial",
                "計算", "複雑", "エントロピー", "情報", "多項式", "シャノン"),
            "探索の困難さがシャノンエントロピー H と多様体不変量 Ξ の分解可能性として定式化される",
            "計算をJonesノットJと見なし検証コスト∫1/(x log x)dx と求解コスト i∫x log x dx は反重力/重力双対の対極にある。" +
                "P=NPは虚residue 1/(2i)≠0 の消滅を要求し矛盾。ゆえに P≠NP。"
        ),
        Problem(
            "hodge", "Hodge Conjecture", "代数幾何・コホモロジー",
            listOf("hodge", "cohomology", "cohom", "dbrane", "brane", "sheaf", "cycle", "projection",
                "ホッジ", "コホモロジー", "代数", "層", "サイクル"),
            "コホモロジー類 cohom D_χ(M) が D-brane のアイソトピーとして多様体上に実現される",
            "∂/∂f F = ∫∫∫ χ(x) cohom D_k(x)[I_m] よりHodge類はD-brane層 cohom D_k の切断で、" +
                "広中の特異点解消とCalabi–Yau分解で各和は代数的サイクル上に台を持つ。係数の有理性はブラケット多項式が Z[A^{±1/4}]⊗Q にあることから従う。"
        ),
        Problem(
            "poincare", "Poincaré Conjecture (解決)", "幾何学的トポロジー",
            listOf("poincare", "thurston", "perelman", "ricci", "flow", "geometriz", "sphere",
                "ポアンカレ", "サーストン", "ペレルマン", "リッチ", "幾何化", "多様体", "トポロジー"),
            "Ricci 流 d/dt g_ij = -2R_ij が大域的多様体を S^3 へ幾何化する（ペレルマンにより解決）",
            "サーストン–ペレルマン分類をエントロピー境界 E(σ)=K(σ)⊗H(σ) で記録し、正規化リッチ流 d/dt g_ij=-2R_ij が" +
                "ペレルマンのW-エントロピーに帰着、単連結性が双曲・Seifert片を除去し有限時間で round S³ に幾何化（解決済み）。"
        ),
        Problem(
            "yang_mills", "Yang–Mills Existence & Mass Gap", "場の量子論",
            listOf("yang", "mills", "gauge", "quantum", "group", "mass", "gap", "quark", "gluon",
                "ヤン", "ミルズ", "ゲージ", "質量", "量子", "クォーク", "場"),
            "ゲージ作用素 ⊕(iℏ∇)^⊕L の質量ギャップが ζ の量子群スペクトルとして現れる",
            "経路積分をHeisenberg流 |ψ(t)⟩=e^{-iĤt}|Ψ⟩ として構成、H_Δ=L(iℏ∇)⊕L=e^{x log x}。" +
                "質量ギャップは虚residue ∬1/(x log x)² dx_m=1/(2i) の非消滅から生じ、Δ(G)>0。"
        ),
        Problem(
            "navier_stokes", "Navier–Stokes Existence & Smoothness", "偏微分方程式・流体",
            listOf("navier", "stokes", "flow", "smooth", "turbulence", "viscous", "gradient", "laplacian",
                "ナビエ", "ストークス", "流体", "滑らか", "乱流", "粘性"),
            "大域的部分微分方程式の滑らかさが Ricci 流型の ∬1/(x log x)² 正則化で保たれる",
            "大域的部分微分方程式の滑らかさは ∬1/(x log x)² 正則化（リッチ流型）で保たれ、渦度の爆発はβ回転shellの" +
                "エネルギー有界性に反するため、滑らかな大域解が存在する。"
        ),
        Problem(
            "bsd", "Birch–Swinnerton-Dyer Conjecture", "数論・楕円曲線",
            listOf("birch", "swinnerton", "dyer", "elliptic", "curve", "lfunction", "rank",
                "バーチ", "楕円曲線", "楕円", "階数", "数論"),
            "楕円曲線の L 関数 L(E,1) が Γ 因子つき β(p,q)/log x ゲージの階数として分解される",
            "楕円曲線のL関数 L(E,s) を s=1 でΓ因子つき β(p,q)/log x ゲージとして展開し、その零点の位数が Mordell–Weil 階数に一致する。"
        )
    )

    private val sigCache = HashMap<String, Double>()
    private fun signatureInvariant(p: Problem): Double =
        sigCache.getOrPut(p.key) { Manifold.xi(p.link) }

    data class Scored(
        val problem: Problem,
        val score: Double,
        val invariant: Double,
        val base: Double,
        val posterior: Double,
        val phase: Double
    )

    data class Analysis(
        val question: String,
        val questionEntropy: Double,
        val questionInvariant: Double,
        val framework: String,
        val dominant: String,
        val dominantResolution: String,
        val posterior: List<Scored>,
        val maxProbDeviation: Double,
        val zerosPreserved: Boolean,
        val unitary: Boolean
    )

    /** Engine-coupled analysis over the seven problems. */
    fun analyze(text: String): Analysis {
        val raw = text.lowercase()
        val tokens = Entropy.tokenize(raw)
        val h = Entropy.shannon(tokens)
        val xiQ = Manifold.xi(tokens)

        data class Pre(val p: Problem, val score: Double, val coupling: Double, val xi: Double)
        val pre = PROBLEMS.map { prob ->
            val overlap = prob.keywords.count { raw.contains(it.lowercase()) }.toDouble()
            val xiP = signatureInvariant(prob)
            val proximity = 1.0 / (1.0 + Math.abs(xiQ - xiP))
            val coupling = Special.zetaGauge(xiQ + 1.0, xiP + 1.0, 2.0 + Math.abs(xiQ - xiP))
            Pre(prob, overlap + proximity * 0.8 + coupling * 0.4, coupling, xiP)
        }

        val base = Engine.softmax(pre.map { it.score })
        val phase = DoubleArray(pre.size) { Math.PI * tanh(pre[it].coupling - 0.5) }
        val mod = Engine.Module("Γ-coupling", base, 0.5, phase)
        val cog = Engine.cognitiveSystem(base, listOf(mod))
        val check = Engine.verify(base, listOf(mod))

        val ranked = pre.mapIndexed { i, x ->
            Scored(x.p, x.score, x.xi, base[i], cog.posterior[i], cog.theta[i])
        }.sortedByDescending { it.posterior }

        return Analysis(
            text, h, xiQ,
            "Y = (M_glob, ζ, β, Γ, J, T, H_Δ) — 大域的微分多様体・ジョーンズ多項式・ゼータ・アカシックTupleSpace",
            ranked.first().problem.name, ranked.first().problem.resolution,
            ranked, check.maxProbDeviation, check.zerosPreserved, check.unitary
        )
    }
}
